[quant] Add a quantized batch_norm operator

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33109 [quant] Add Quantized BatchNorm2d module
* **#33080 [quant] Add a quantized batch_norm operator**

Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19861927](https://our.internmc.facebook.com/intern/diff/D19861927)

supriyar committed:
[quant] Add a quantized batch_norm operator
Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add a quantized batch_norm operator"


Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add a quantized batch_norm operator"


Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -476,6 +476,11 @@
 
 - func: batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float momentum, float eps, bool cudnn_enabled) -> Tensor
 
+- func: quantized_batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor mean, Tensor var, float eps, float output_scale, int output_zero_point) -> Tensor
```
We should also register the variant with the relu (quantized_batch_norm_relu), like other cases:https://github.com/pytorch/pytorch/blob/ecbf6f99e6a4e373105133b31534c9fb50f2acca/aten/src/ATen/native/quantized/cpu/qconv.cpp#L640

supriyar replied:
Let's leave it at quantized batch_norm support for now and add relu fusion later (if required)

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qbatch_norm.cpp:
```diff
@@ -0,0 +1,167 @@
+  //         + bias(c)
+  // We factor out inv_var(c) = 1 / sqrt(var(c) + eps).
+  for (int64_t c = 0; c < channels; c++) {
+    float inv_var = 1 / std::sqrt(var_data[c] + static_cast<float>(eps));
```
Lets call in inv_sigma as this is the standard deviation, not the variance.

jamesr66a replied:
1.0f

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -1071,6 +1071,32 @@ def equal_ref(qX, qX2):
         self.assertEqual(qX.equal(qX), equal_ref(qX, qX))
         self.assertEqual(qX.equal(qX2), equal_ref(qX, qX2))
 
+    @given(X=hu.tensor(shapes=hu.array_shapes(min_dims=4, max_dims=4,
```
Can we also test for qx being of type qint8?

supriyar replied:
Added an assertion for type other than uint8. Updated test to use both qengines.

I'm not sure how we can make sure to test the AVX2 and default path as it is guarded by compile time macro. I am relying on CI to run on machines without AVX2.

raghuramank100 reviewed (changes requested):
Looks good! A few requests:

jamesr66a commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -955,6 +961,126 @@ void qtopk_kernel(Tensor& values,
+    const float* alpha,
+    const float* beta,
+    uint8_t* Y) {
+#if defined(__AVX2__)
```
```
#if defined(__AVX2__) && defined(__FMA__)
```

jamesr66a commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -955,6 +961,126 @@ void qtopk_kernel(Tensor& values,
+void q_batch_norm_kernel(
+    int N,
+    int C,
+    int HxW,
```
We sure we want to use `int` for all these shape parameters? It's probably better if this is at least fixed-width, and preferrably `int64_t` to match the rest of the index handling in pt

jamesr66a commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -955,6 +961,126 @@ void qtopk_kernel(Tensor& values,
+    int C,
+    int HxW,
+    const int in_zero_point,
+    const int out_zero_point,
```
zero points should definitely be `int64_t`, right?

supriyar replied:
Yeah, updated them

jamesr66a commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -955,6 +961,126 @@ void qtopk_kernel(Tensor& values,
+    int r = C % kVLen;
+    const uint8_t* X_ptr = X + i * C;
+    uint8_t* Y_ptr = Y + i * C;
+    for (int j = 0; j < n; j += kVLen) {
```
Can this instead be done 4x interleaved to ensure better pipeline occupancy? You can then replace most of these instructions with calls to Vec256<quint8>::quantize (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vec256/vec256_qint.h#L436)

That also cleans up the unnecessary shuffle instructions etc

jamesr66a replied:
NB: I haven't compiled or run this so there might be some details off! :p

```
using Vec = Vec256<quint8>;

// Hoisted variables
auto in_zp_vec = Vec256<float>(static_cast<float>(in_zero_point));
auto fake_scale = Vec256<float>(1.0f);
auto scale_neg_zp_premul = fake_input_scale + in_zp.vec.neg();
auto out_zero_point_v = Vec(c10::quint8(out_zero_point));

for (int64_t j = 0; j < n; j += Vec::float_num_vecs()) {
  auto vals_q = Vec::loadu(X_ptr + j);
  // Fake scale of 1.0 here, should not affect performance (FMA in place of sub)
  auto vals_dq = vals_q.dequantize(fake_in_scale, in_zp_vec, scale_neg_zp_premul);
  for (int idx = 0; idx < vals_dq.size(); ++idx) {
    auto alpha_v = Vec256<float>(alpha + j * Vec::float_num_vecs() + idx);
    auto beta_v = Vec256<float>(beta + j * Vec::float_num_vecs() + idx);
    vals_dq[idx] = vec256::fmadd(vals_dq[idx], alpha_v, beta_v);
  }
  // Fake scale again
  auto outputs_q = Vec::quantize(vals_dq, /*output_scale=*/1.0f, out_zero_point, /*inv_output_scale=*/1.0f);
  if (ReluFused) {
    outputs_q = outputs_q.relu(out_zero_point_v);
  }
  outputs_q.store(Y_ptr + j);
}
```

jamesr66a commented on aten/src/ATen/native/quantized/cpu/qbatch_norm.cpp:
```diff
@@ -0,0 +1,167 @@
+  const int N = qx.size(0);
+  const int C = qx.size(1);
+  const int H = qx.size(2);
+  const int W = qx.size(3);
```
All of these are narrowing conversions `int` -> `int64_t`

supriyar committed:
Update on "[quant] Add a quantized batch_norm operator"


Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add a quantized batch_norm operator"


Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add a quantized batch_norm operator"


Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add a quantized batch_norm operator"


Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jamesr66a commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -955,6 +961,89 @@ void qtopk_kernel(Tensor& values,
+  constexpr int kVLen = 8;
+  const int outer_size = N * HxW;
+  using Vec = Vec256<quint8>;
+  for (int i = 0; i < outer_size; ++i) {
```
I think we might be able to simplify this further, since--correct me if i'm wrong--this is essentially just an elementwise access pattern where we broadcast `alpha` and `beta` to all [N, H, W]. It might be pretty simple to just use TensorIterator instead of handling the loop nest explicitly. I can elaborate a bit more later

supriyar replied:
Blocked by #33166

supriyar committed:
Update on "[quant] Add a quantized batch_norm operator"


Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add a quantized batch_norm operator"


Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jamesr66a commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -955,6 +961,70 @@ void qtopk_kernel(Tensor& values,
+
+  constexpr int kVLen = 8;
+  const int64_t outer_size = N * HxW;
+  using Vec = Vec256<quint8>;
```
We can generalize this with AT_FORALL_QINT_TYPES now that we've gotten rid of the bare intrinsics

jamesr66a reviewed (approved):
LG, just one comment

supriyar committed:
Update on "[quant] Add a quantized batch_norm operator"


Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19861927](https://our.internmc.facebook.com/intern/diff/D19861927)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add a quantized batch_norm operator"


Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19861927](https://our.internmc.facebook.com/intern/diff/D19861927)

[ghstack-poisoned]
